app/uk/gov/hmrc/agentmapping/auth/AuthActions.scala

```diff
@@ -63,6 +63,15 @@ class AuthActions @Inject()(val authConnector: AuthConnector) extends Authorised
       .recover { handleError }
   }
 
+  def AuthorisedWithProviderId(body: Request[AnyContent] => String => Future[Result]): Action[AnyContent] =
+    Action.async { implicit request =>
+      authorised(AuthProviders(GovernmentGateway) and AffinityGroup.Agent)
+        .retrieve(Retrievals.credentials) {
+          case credentials => body(request)(credentials.providerId)
+          case _           => Future.successful(Forbidden)
+        }
+    }
+
```

Comment:
you can add ` .recover { handleError }` block here too (same as above method )? 
but it needs a little clean up as `handleError` do not need to be passed from calling method

look at subscription frontend logic :

```  private def handleException(implicit ec: ExecutionContext, request: Request[_]): PartialFunction[Throwable, Result] = {

    case _: UnsupportedAffinityGroup =>
      mark("Count-Subscription-NonAgent")
      Redirect(routes.StartController.showNotAgent())

    case _: NoActiveSession =>
      toGGLogin(if (appConfig.isDevMode) s"http://${request.host}${request.uri}" else s"${request.uri}")

    case _: UnsupportedAuthProvider =>
      Logger.warn("User is not logged in via  GovernmentGateway, signing out and redirecting")
      Redirect(routes.SignedOutController.signOut())
  }```